src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64VaList.java

```diff
@@ -0,0 +1,260 @@
+        return EMPTY;
+    }
+
+    @Override
+    public int nextVarg(ValueLayout.OfInt layout) {
+        return (int) read(layout);
+    }
+
+    @Override
+    public long nextVarg(ValueLayout.OfLong layout) {
+        return (long) read(layout);
+    }
+
+    @Override
+    public double nextVarg(ValueLayout.OfDouble layout) {
+        return (double) read(layout);
+    }
+
+    @Override
+    public MemorySegment nextVarg(ValueLayout.OfAddress layout) {
+        return (MemorySegment) read(layout);
+    }
+
+    @Override
+    public MemorySegment nextVarg(GroupLayout layout, SegmentAllocator allocator) {
+        Objects.requireNonNull(allocator);
+        return (MemorySegment) read(layout, allocator);
+    }
+
+    private Object read(MemoryLayout layout) {
+        return read(layout, SharedUtils.THROWING_ALLOCATOR);
+    }
+
+    private Object read(MemoryLayout layout, SegmentAllocator allocator) {
+        Objects.requireNonNull(layout);
+        Object res;
+        if (layout instanceof GroupLayout) {
+            TypeClass typeClass = TypeClass.classifyLayout(layout);
+            res = switch (typeClass) {
+                case STRUCT_REFERENCE -> {
+                    checkElement(layout, VA_SLOT_SIZE_BYTES);
+                    MemorySegment structAddr = (MemorySegment) VH_address.get(segment);
+                    MemorySegment struct = MemorySegment.ofAddress(structAddr.address(), layout.byteSize(), segment.session());
+                    MemorySegment seg = allocator.allocate(layout);
+                    seg.copyFrom(struct);
+                    segment = segment.asSlice(VA_SLOT_SIZE_BYTES);
+                    yield seg;
+                }
+                case STRUCT_REGISTER, STRUCT_HFA -> {
+                    long size = alignUp(layout.byteSize(), VA_SLOT_SIZE_BYTES);
+                    checkElement(layout, size);
+                    MemorySegment struct = allocator.allocate(layout)
+                            .copyFrom(segment.asSlice(0, layout.byteSize()));
+                    segment = segment.asSlice(size);
+                    yield struct;
+                }
+                default -> throw new IllegalStateException("Unexpected TypeClass: " + typeClass);
+            };
+        } else {
+            checkElement(layout, VA_SLOT_SIZE_BYTES);
+            VarHandle reader = layout.varHandle();
+            res = reader.get(segment);
+            segment = segment.asSlice(VA_SLOT_SIZE_BYTES);
+        }
+        return res;
+    }
+
+    private static long sizeOf(MemoryLayout layout) {
+        return switch (TypeClass.classifyLayout(layout)) {
+            case STRUCT_REGISTER, STRUCT_HFA -> alignUp(layout.byteSize(), VA_SLOT_SIZE_BYTES);
+            default -> VA_SLOT_SIZE_BYTES;
+        };
+    }
+
+    @Override
+    public void skip(MemoryLayout... layouts) {
+        Objects.requireNonNull(layouts);
+        MemorySessionImpl.toSessionImpl(segment.session()).checkValidState();
```

Comment:
Calls to `toSessionImpl` can be replaced with simple cast
```suggestion
        ((MemorySessionImpl) segment.session()).checkValidState();
```